<a href="https://colab.research.google.com/github/kezamamio/58025_KFCBreak_RevisitingPython/blob/main/58025_KFCBreak_RevisitingPython.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Welcome to Python Fundamentals
In this module, we are going to establish or review our skills in Python programming. In this notebook we are going to cover:
* Variables and Data Types 
* Operations
* Input and Output Operations
* Logic Control
* Iterables
* Functions

## Variable and Data Types

This section introduces how to define certain variables and classify their data types.

In [ ]:
x = 1
a,b = 0, -1
print(b, a, x)

-1 0 1


The line of codes above represents the basic initialization of variables with an ***int*** data type. To print the values, simply utilize the ***print*** statement with the variable inside the parenthesis.

In [ ]:
print(type(x))

<class 'int'>


The line of code above represents how to classify a specific variable's data type. By using the ***print*** statement, it can display the defined variable's data type. Since the previous code snippet defined ***x*** as an integer, the next code snippet can read the defined variable.

In [ ]:
y = 1.0
print(y)
print(type(y))

1.0
<class 'float'>


The line of codes above represents an initialization of a float data type. By using a ***print*** statement, the defined variable and data type are displayed.

In [ ]:
x = float(x)
print(type(x))

<class 'float'>


The line of codes above represents the conversion of a certain data type into another. From the previous code snippets, the declaration of ***x*** was previously defined as an ***int*** data type. By utilizing this syntax, the previous data type is now converted into ***float***.

In [ ]:
s,t,u = "0", '1', 'one'
print(f"The variable s is defined as {s} and has a data type of {type(s)}.")
print(f"The variable t is defined as {t} and has a data type of {type(t)}.")
print(f"The variable u is defined as {u} and has a data type of {type(u)}.")

The variable s is defined as 0 and has a data type of <class 'str'>.
The variable t is defined as 1 and has a data type of <class 'str'>.
The variable u is defined as one and has a data type of <class 'str'>.


The line of codes above represents the declaration of strings. Strings, referred to as ***str***, is defined by two double quotations. Since the variable's values are enclosed in double quotations, they are read as an ***str*** data type.

In [ ]:
s_int = int(s)
print(f"{s_int} is converted into {type(s)}.")

0 is converted into <class 'int'>.


The line of codes above represents the conversion of an ***str*** data type into an ***int*** data type. If the value inside the string is an integer, it is valid to convert into an ***int***. Otherwise, it will return an error.

## Operations
Python operations are utilized to perform procedure on variables and its values. It features a wide variety of operations, similar to any other programming language. Python can execute operations such as **arithmetic**, **assignment**, **comparators**, **logical**, and **input/output** operations. 

### Arithmetic
Arithmetic operators are used to perform fundamental mathematical operations between two numeric values.

In [ ]:
a,b,c,d = 2.0, -0.5, 0, -32

Above is the line of code used to initialize the numeric values.

In [ ]:
### Addition1
S = a+b
S

1.5

In [ ]:
### Addition2
print (a + b)

1.5


The line of codes above shows how to perform addition operation in Python with two different ways. In **Addition1**, the initialized addends ***float*** values **a**  and **b** are stored in variable **S**. While in **Addition2**, the output operation ***print()*** is utilized to execute the given operation. Moreover, the sum for both codes were still the same. 

In [ ]:
### Subtraction1
D = b-d
D

31.5

In [ ]:
### Subtraction2
print (b - d)

31.5


The following line of codes above illustrated how to perform subtraction operation in Python. **Subtraction1** stored the ***float***  values **b** (minuend) and **d** (subtrahend) first in variable **D** then execute it. While **Subtraction2** used ***print()*** output operation. Both codes displayed the same result of 31.5.



In [ ]:
### Multiplication1
P = a*d
P

-64.0

In [ ]:
### Multiplication2
print (a*d)

-64.0


The line of codes above shows how to perform multiplication operation using Python. **Multiplication1** stored the ***float*** values **b** (multiplicand) and **d** (multiplier) first in variable **P** then execute it. While **Multiplication2** used ***print()*** output operation. Nevertheless, both codes displayed the same product.

In [ ]:
### Division
Q = c/a
Q

0.0

In [ ]:
### Division
print (c/a)

0.0


Above were the line of codes that illustrates how to perform division operation using Python. **Division1** stored the ***float*** values **c** (dividend) and **a** (divisor) first in variable **Q** then execute it. While **Division2** used ***print()*** output operation. Both codes still displayed the same quotient output.

In [ ]:
### Floor Division
Fq = a//d
Fq

-1.0

Simply put, floor division **rounds the quotient down to the nearest integer**, thus returning to the floor of the result. The real quotient of the sample code above was -0.0625. Because of floor division, it was rounded down to -1.0. 

In [ ]:
### Exponentiation1
E = a**d
E

2.3283064365386963e-10

In [ ]:
### Exponentiation2
print (a**d)

2.3283064365386963e-10


Exponentiation operation pertains to performing calculations that uses exponent or power on its operators. It was characterized by the symbol **. Exponentiation1** code above illustrates how to perform this operation by storing the declared values into variable named as **E**. **Exponentiation2** is a much simpler coding style where ***print()*** was utilized.

In [ ]:
### Modulo1
mod = b%d
mod

-0.5

In [ ]:
###M Modulo2
print (b%d)

-0.5


Modulo operator returns the remainder as a result of dividing two values. In the sample code, **Modulo1** utilized variable storage for the operation, while in **Modulo2**, ***print()*** was used. Both have the same resulting output. 

### Assignment Operations
Assignment operators are used to designate or store a value into a variable.

In [ ]:
G, H, J, K, F, C = 0, 100, 150, 5, 3, 21

In [ ]:
G += a
G

2.0

In the line code above, both left and right operands ***G*** and ***a*** respectively were added to each other. The resulting sum was then assigned to left operand G. Simply put, the equivalent expression would be **G = G + a**. 

In [ ]:
H -= c
H

100

In the line code above, both left and right operands ***H*** and ***c*** respectively were subtracted to each other. The resulting difference was then assigned to left operand ***H***. Simply put, the equivalent expression would be **H = H - c**. The ***c*** in this expression was derived from the value of c stored in the initialized values in the arithmetic operations.

In [ ]:
J *= 2
J

300

In the line code above, both left and right operands ***J*** and ***2*** respectively were multiplied to each other. The product was then assigned to left operand ***J***. Simply put, the equivalent expression would be **J  = J * 2**. It was also noted that whenever we run the code, the previous product will be multiplied by 2. By continuous running, the product becomes relatively of a larger value. This case was also similar if we continuously run the assignment operators that uses addition, and subtraction. 

In [ ]:
K **= 2
K

25

In the line code above, both left and right operands ***K*** and ***2*** utilizes exponential operation. The output was then assigned to left operand ***K***. Simply put, the equivalent expression would be **K  = K ** 2**. 

In [ ]:
F %= 3
F

0

In the line code above, both left and right operands ***F*** and ***3*** used modulo operation. The output was then assigned to left operand ***F***. The equivalent expression would be **F  = F & 3** or **F = 5 % 3**. 

In [ ]:
C /= 3
C

7.0

In the line code above, both left and right operands ***C*** and ***3*** used division operation. The quotient was then assigned to left operand ***C***. The equivalent expression would be **C  = C / 3** or **C = 21 / 3**. 

### Comparators
Comparison operators also called as comparators are used to compare two given values based on their magnitude of equality.

In [ ]:
res_1, res_2, res_3 = 1, 2.0, 0.5
true_val = 1.0

Above is the line of code used to initialize the values for each variables.

In [ ]:
## Equality
res_1 == true_val

True

In the code above, **res_1** has a value of **1** while **true_value** variable has also a value of **1**. Therefore, the two operands are equal and the resulting output is **True**. 

In [ ]:
## Non-equality
res_2 != true_val

True

In the code above, **res_2** has a ***float*** value of **2.0** while **true_value** variable has a value of **1**. These two operands are not equal therefore the resulting output will be **True**. 

In [ ]:
## Greater than or equal
res_1 >= res_2

False

**res_2** and **res_1** are evidently not equal to one another and that **res_1** is not also greater than **res_2**. Therefore, the output would be **False**.

In [ ]:
### Less than or equal
res_2 <= res_3

False

**res_2** and **res_3** are not equal to one another and that **res_2** that has a value of 2.0 is not less than **res_3**'s value of 0.5. Therefore, the output would be **False**.

In [ ]:
## Inequality
t1 = res_1 > res_2
t2 = res_1 < res_2/2
t3 = res_1 >= res_2/2
t4 = res_1 <= res_2  
t1

False

The code above illustrates inequality operation. The series of codes will evaluate conditions applied between **res_1** and **res_2**. 

### Logical
Logical operators are mainly used for synthesizing conditional statements.


In [ ]:
u = 15
u > 3 and u < 10

False

The code above returned an output display of ***False*** because 15 is greater than 3 (True) but 15 is not less than 10 (False). Statements that are evaluated using **and** operators should return **True** if both statements or conditions are true. 

In [ ]:
p, q = True, False
conj = p and q
conj

False

In [ ]:
f = 5
f <= 5 or f < 4

True

The code above returned an output display of ***True*** even if 5 is not less than 4 (False). Statements that are evaluated using **or** operators should return **True** if one of the statements or conditions are true. In this case, the first statement was satisfied. 

In [ ]:
p, q = True, False
disj = p or q
disj

In [ ]:
res_1 == true_val

True

**res_1** that has a value of 1 wass indeed equal to variable **true_val**'s value. Thus, output displayed was **True**.

In [ ]:
res_1 is true_val

False

But, **res_1** will never not be **true_val** even if they have the same value. This line of code uses the **is** logical operator. It will only return an output of **True** if the variables on either side of the operator describes the same variable (***true_val is true_val***). Otherwise, it is considered **False**.

In [ ]:
res_1 is not true_val

True

Contrary to **is** operator, **is not** returns display **True** if the variables on either side are not pertaining to the **same** object. Otherwise, the output will be **False**.

In [ ]:
p, q = True, False
nand = not(p and q)
nand

True

The line of code above describes a statement for NAND expression. It is a combination of NOT and AND operators. Upon running, the code will first evaluate the expression inside the parenthesis based on operator precedence. Since it uses an AND operator, it will return **False**. After which, NOT False is evaluated. Therefore, the output is **True**.  

In [ ]:
p, q = True, False
xor = (not p and q) or (p and not q)
xor

True

The line of code above describes a statement for an XOR expression. Upon running, the code will first evaluate the expression inside the parentheses based on operator precedence. First expression will result to **False** while second parenthesis is evaluated with a result of **True**. Since it uses an OR operator, the final output will thereby return **True**. 

### I/O
Input operators allows users to enter data. Whereas, output operator such as ***print()*** is used to display results processed by the Python program. 

In [ ]:
print("Hello World")

Hello World


The code above will display the output **Hello World** because of the output operation ***print()***.

In [ ]:
cnt = 1

In [ ]:
string = "Hello World"
print(string, ", Current run count is:", cnt)
cnt += 1

Hello World , Current run count is: 3


The main goal of the code above was to display a string output and the current run count (or how many times the code has been ran). The variable **string** stored **Hello World** with a ***str*** data type. More so, **cnt** variable was also initialized with a value of 1. To determine the current count, assignment operation of addition was used. 

In [ ]:
print(f"{string}, Current run count is: {cnt}")

Hello World, Current run count is: 4


The code above has the same output display from the previous code but this time, ***f string*** was used. This is another way to format a string syntax. 

In [ ]:
sem_grade = 82.243564657461234
name = ""
print("Hello {}, your semestral grade is: {}".format(name, sem_grade))

Hello , your semestral grade is: 82.24356465746123


Above is a line code that should display a semestral grade. Similar to previous codes, variable **sem_grade** was first initialized with a float dtata type value then displayed it using ***print()*** statement. Along with it is a ***format ***code to specify which values should be displayed first. 

In [ ]:
w_pg, w_mg, w_fg = 0.3, 0.3, 0.4
print("The weights of your semestral grades are:\
\n\t{:.2%} for Prelims\
\n\t{:.2%} for Midterms, and\
\n\t{:.2%} for Finals.".format(w_pg, w_mg, w_fg))

The weights of your semestral grades are:
	30.00% for Prelims
	30.00% for Midterms, and
	40.00% for Finals.


Above is a line code that displays the percentage weight of each periodic terms to accumulate the semestral grade. Similar to previous codes, ***format*** code was also used to specify which values should be displayed first. Also, prelim, midterm, and final grades values were initialized and stored through its corresponding variables. To display the output, ***print()*** is used. 

In [ ]:
x = input("enter a number: ")
x

enter a number: 1


'1'

In the code above, the program prompts the user to enter a number. After which, the number inputted was displayed. This was mainly because of the ***input()*** statement utilized in the code. 

In [ ]:
name = input("Enter your name bhie: ")
pg = input("Enter prelim grade: ")
mg = input("Enter midterm grade: ")
fg = input("Enter finals grade: ")
sem_grade = None
print("Hello {}, your semestral grade is: {}".format(name, sem_grade))

Enter your name bhie: Keziah
Enter prelim grade: 90
Enter midterm grade: 89
Enter finals grade: 75
Hello Keziah, your semestral grade is: None


The code above is a more complex one compared to previous codes. Basically, the user was prompted to enter data such as name, prelim, midterm, and finals grade. After which, it will display the semestral grade which is equal to none since no formula was declared to compute for the ***sem_grade***. 

# Looping Statements

## While

In [ ]:
## while loops
i, j = 0, 10
while(i<=j):
    print(f"{i}\t|\t{j}")
    i+=1

0	|	10
1	|	10
2	|	10
3	|	10
4	|	10
5	|	10
6	|	10
7	|	10
8	|	10
9	|	10
10	|	10


The code above uses while loop to increment the value of **i** by 1 until it is equivalent to the value of **j**. Since the initial value of **i** is zero the loop was executed 11 times for **i** to become 10.

## For

In [ ]:
# for(int i=0; i<10; i++){
# printf(i)
# }

i=0
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


The code above uses *for* loop to increment the value of **i** until the values of **i** was equivalent to the range. Since the initial value of **i** is 0 the loop increment the value of **i** for the number of times is in *range()*

In [ ]:
playlist = ['Blessing Chord', 'ZEAL of proud']
print('Now Playing:\n')
for song in playlist:
    print(song)

Now Playing:

Blessing Chord
ZEAL of proud


The code above uses for loop to print all values inside **playlist**. The loop automatically terminates once there are no values present inside **playlist**.

# Flow Control

## Condition Statements

In [ ]:
numeral1, numeral2 = 12, 12
if(numeral1 == numeral2):
    print("Yey")
elif(numeral1>numeral2):
    print("Hoho")
else:
    print("Aww")
print("Hip hip")

Yey
Hip hip


The code above have 2 variables **numeral1** and **numeral2**. The code checks if their values are equal, if **numeral1** is greater than **numeral2**, or if does not fit the the other conditions. If both values are equal it satisfies the first condition and will print "Yey", if the not it will check the second condition and will print "Hoho" if it satisfy the condition, and will print "Aww" if neither of the conditions are satisfied.

# Functions
This section utilizes the use of a function. A function is a block of code that will only run when called upon.In Python, you can create a function by using the keyword ***def***. The data inside it are referred to as ***parameters***, while the information that can be passed into functions are called as ***arguments***.


In [ ]:
# void DeleteUser(int userid){
#     delete(userid);
# }

def delete_user (userid):
    print("Successfully deleted user: {}".format(userid))
    
def delete_all_users ():
    print("Successfully deleted all users.")

delete_all_users()

Successfully deleted all users.


The above code had two functions: ***delete_user()*** and ***delete_all_users()***. Notice that while both had the statements to execute printing, the statement for the ***delete_all_users()*** is the only one outputted because it is called upon while the other is not.

In [ ]:
userid = 0

delete_user(0)
delete_all_users()


Successfully deleted user: 201812893
Successfully deleted all users.


In this block of code, the statements inside  both ***delete_user()*** and ***delete_all_users()*** are ouputted because both functions are called upon. Otherwise, there would be no statement that will be appear.

In [ ]:
def add(addend1, addend2):
    print('The sum of', addend1, 'and', addend2, 'is :')
    return addend1 + addend2

def power_of_base2(exponent):
    print('Two to the power or', exponent, 'is :')
    return 2**exponent

add(2,3)

The sum of 2 and 3 is :


5

Observing from the above code, a function is not only limited to passing printed stataments but also, pass printed statements and return an equation by taking arguments. ***Arguments*** can be put inside the parenthesis beside the function name. Multiple arguments are allowed, provided that a comma separates them.


### Lambda Functions
A lambda function is a small anonymous function. It is useful especially when one requires a nameless function for a short period of time. This type of function can have multiplt arguments but only one expression. It has a syntax of **lambda *arguments* : expression**.

In [ ]:
x = 4

In [ ]:
def f(x):
  return 2*(x*x)-1
  f(x)

31

This block uses a standard function - with the **function f(x)** calls whatever the value of x is, and applies it to the parameters inside - in this case, the value of x is already determined as 4. Hence, when f(x) is invoked, it takes out the value of x, processes the return statement, and prints out an answer.

In [ ]:
g = lambda x: 2*(x*x)-1
print(g(x))

31


This block of code is similar to the above standard function but with the difference of having used a lambda function. The function here is **g** while **x** is the argument, and **2*(x*x)-1** is the expression. Compared to the standard function, the way of the writing of the code had been relatively shorter.

# Semestral Grade Calculator

This section introduces the group's code implementation of a grade calculator that will accept the user's name, program, and course and compute their semestral grade based on their term grades.

In [ ]:
# -*- coding: utf-8 -*-
"""
Created on Thu Feb 11 07:19:59 2021

@author: KFC Break
Semestral Grade Calculator
"""

happy, lol, sad = "\U0001F600","\U0001F923","\U0001F619"
print('Good day, student! Please enter the required information.')
name = input('Name: ')
program = input('Program: ')
course = input('Course: ')
print(f"{name}, kindly input your grades per term for the course {course}.")
p = float(input('Prelim Term: '))
m = float(input('Midterm: '))
f = float(input('Finals Term: '))
SG = (p*0.3)+(m*0.3)+(f*0.4)
SG2 = "{:.2f}".format(SG)      # To print grade with 2 decimal places
if SG > 70.00:
    print(f"Your semestral grade for the course {course} is {SG2} {happy} ")
elif SG == 70.00:
    print(f"Your semestral grade for the course {course} is {SG2} {lol} ")
elif SG < 70.00:
    print(f"Your semestral grade for the course {course} is {SG2} {sad} ")
else:
    print("It seems you're not taking this calculator seriously. Please try again.")

Good day, student! Please enter the required information.
Name: Jene Reiner N. Borejon
Program: BS Computer Engineering
Course: Fund. of AI and Data Analytic
Jene Reiner N. Borejon, kindly input your grades per term for the course Fund. of AI and Data Analytic.
Prelim Term: 98
Midterm: 97
Finals Term: 98
Your semestral grade for the course Fund. of AI and Data Analytic is 97.70 😀 
